gobblin-utility/src/main/java/gobblin/broker/SharedResourcesBrokerImpl.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (C) 2014-2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.broker;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ExecutionException;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+import gobblin.broker.iface.NoSuchScopeException;
+import gobblin.broker.iface.NotConfiguredException;
+import gobblin.broker.iface.ScopeInstance;
+import gobblin.broker.iface.ScopeType;
+import gobblin.broker.iface.SharedResourceFactory;
+import gobblin.broker.iface.SharedResourceKey;
+import gobblin.broker.iface.SharedResourcesBroker;
+import gobblin.util.ConfigUtils;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.Data;
+
+
+/**
+ * An implementation of {@link SharedResourcesBroker} using a {@link DefaultBrokerCache} for storing shared objects.
+ *
+ * Instances of this broker must be created as follows:
+ * <pre>
+ *  SharedResourcesBrokerImpl<MyScopes> rootBroker = SharedResourcesBrokerFactory.createDefaultTopLevelBroker(myConfig);
+ *  SharedResourcesBrokerImpl<MyScopes> scopeBroker = topBroker.newSubscopedBuilder(scope, "scopeId").build();
+ * </pre>
+ */
+public class SharedResourcesBrokerImpl<S extends ScopeType<S>> implements SharedResourcesBroker<S> {
+
+  private final DefaultBrokerCache<S> brokerCache;
+  private final ScopeImpl<S> leafScope;
```

Comment:
nit: i find the naming slightly confusing since there is only one leaf. How about "selfScope" ?